

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>CSVCC &mdash; GOR Open Source v.2.0 documentation</title>
  

  
  
    <link rel="shortcut icon" href="../_static/favicon.ico"/>
  
  
  

  
  <script type="text/javascript" src="../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/language_data.js"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../_static/css/gorpipe.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <nav class="wy-master-nav">
    <div class="gor-logo-home">
        <a href="../index.html">
            <img src="../_static/GORLogoText.svg" class="logo" />
        </a>
    </div>
    <div role="search" class="wxnc-search-form">
        <button class="wxnc-search-button" form="wxnc-search-form" formmethod="get">
            <i class="fa fa-search" aria-hidden="true"> </i>
        </button>
        <form id="wxnc-search-form" class="wy-form" action="../search.html" method="get">
            <input class="wxnc-no-right-border" size="30" type="text" name="q" placeholder="Search the user manual..."/>
            <input type="hidden" name="check_keywords" value="yes" />
            <input type="hidden" name="area" value="default" />
        </form>
    </div>
</nav>

  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            
            
              <div class="version">
                <a href="../index.html">
                  GOR Open Source
                  <p class="wxnc-subtitle">Documentation</p>
                </a>
              </div>
            
          

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../introduction.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="../commands.html">Commands</a></li>
<li class="toctree-l1"><a class="reference internal" href="../functions.html">Functions</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">GOR Open Source</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html">Docs</a> &raquo;</li>
        
      <li>CSVCC</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <span style="float:right; padding-top:10px; color:#BABABA;">Used in: gor only</span><div class="section" id="csvcc">
<span id="id1"></span><h1>CSVCC<a class="headerlink" href="#csvcc" title="Permalink to this headline">¶</a></h1>
<p>The <strong>CSVCC</strong> command is used aggregate or count genotypes stored in horizontal values column in a CSV or fixed element size format.
See <a class="reference internal" href="CSVSEL.html#csvsel"><span class="std std-ref">CSVSEL</span></a> for more description of the requirements for the data input.</p>
<p>The aggregation is grouped according to the phenotype relation, which can be of the form (tag,status) or (tag,pheno,status).
Typically, the relation is a binary relation with PN and case vs control status.  Affection status for multiple phenotypes
can be represented by using the class.</p>
<p>Note, only non-zero GTcounts are in the output, hence when this output is used with <a class="reference internal" href="PIVOT.html#pivot"><span class="std std-ref">PIVOT</span></a> the -e option should be used with zero,
e.g. see examples below.</p>
<p>Also, CSVCC is slower when GT values outside the range 0-6 are used and when a separator is used instead of fixed value size.</p>
<div class="section" id="usage">
<h2>Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h2>
<div class="highlight-gor notranslate"><div class="highlight"><pre><span></span><span class="nd">gor</span> ... | <span class="nb">CSVCC</span> tagbucketrelation phenorelation [-<span class="no">gc</span> cols] [-u undefsymbol] [-s sep] [-vs charsize]
</pre></div>
</div>
</div>
<div class="section" id="options">
<h2>Options<a class="headerlink" href="#options" title="Permalink to this headline">¶</a></h2>
<table class="docutils align-default">
<colgroup>
<col style="width: 17%" />
<col style="width: 83%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">-gc</span> <span class="pre">cols</span></code></p></td>
<td><p>Grouping columns other than (chrom,pos)</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">-s</span> <span class="pre">sep</span></code></p></td>
<td><p>Specify separator for the elements in values.  Default to comma.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">-vs</span> <span class="pre">size</span></code></p></td>
<td><p>Use a fixed character size for values, e.g. rather than variable length separated with comma.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">-u</span> <span class="pre">undefsymbol</span></code></p></td>
<td><p>The symbol to use in case a bucket does not have the corresponding position and grouping columns.
Default value is 4.  Typical values are 3, 4 or NA. Defaults to 3.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">-probunphased</span></code></p></td>
<td><p>The input is probabilistic genotypes, two chars, representing Pr(gt=1) and Pr(gt=2).
The probability is Phred-like, defining probabilities from 1(!) to 0(~) as round((1.0-pr)*93.0)+33
With this option the GTcount column is floating point number, unless threshold is used.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">-probphased</span></code></p></td>
<td><p>The input is probabilistic haplotypes, two chars, representing Pr(father=Alt) and Pr(mother=Alt
Chars defined as in unphased.
Pr(gt=1) = Pr(father=alt)*(1-Pr(mother=Alt))+Pr(mother=alt)*(1-Pr(father=Alt)) etc
With this option the GTcount column is floating point number, unless threshold is used.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">-threshold</span> <span class="pre">val</span></code></p></td>
<td><p>A threshold to discretise the genotypes.  Typical values in the range 0.75-1.0
This causes gt=3 to be used if the Pr(gt=n) &lt; threshold, for all n in 0,1,2</p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="examples">
<h2>Examples<a class="headerlink" href="#examples" title="Permalink to this headline">¶</a></h2>
<p>With <cite>[#PNbuckets#]</cite> and <cite>[#horvars#]</cite> defined analogous manner as in <a class="reference internal" href="CSVSEL.html"><span class="doc">CSVSEL</span></a>, we could perform case-control analysis as</p>
<div class="highlight-gor notranslate"><div class="highlight"><pre><span></span>/* Creating arbitrary case control status on PNs */

<span class="nd">create</span> #casecontrol# = <span class="nd">nor</span> [#PNbuckets#] | select #<span class="na">PN</span> | calc ccstatus = if(random()&gt;0.5,&#39;CASE&#39;,&#39;CTRL&#39;);

<span class="nd">gor</span> [#horvars#] | <span class="nb">CSVCC</span> [#PNbuckets#] [#casecontrol#] -<span class="no">gc</span> #3,#4 -u 3
| <span class="nb">PIVOT</span> -<span class="no">gc</span> <span class="na">reference</span>,<span class="na">call</span>,ccstatus GT -v 0,1,2,3 -e 0
| <span class="nb">PIVOT</span> -<span class="no">gc</span> <span class="na">reference</span>,<span class="na">call</span> -v CASE,CTRL ccstatus -e 0
| <span class="nb">CALC</span> pVal_mm EFORM(PVAL(case_2_GTcount*2+case_1_GTcount,case_0_GTcount*2
                          + case_1_GTcount,ctrl_2_GTcount*2+ctrl_1_GTcount,ctrl_0_GTcount*2 + ctrl_1_GTcount),5,1)
</pre></div>
</div>
<p>The above query uses <a class="reference internal" href="CSVSEL.html"><span class="doc">CSVSEL</span></a> to count the number of different allele states; 0 (homozygous reference), 1 (heterozygous), 2 (homozygous) and 3 (unknown)
per variant, e.g. (chrom,pos,reference,call).  Using <code class="docutils literal notranslate"><span class="pre">-u</span> <span class="pre">3</span></code> means that absence of data is treated as a “stored unknown”.  By using a
value such as 0, one can for instance assume that absence of data means homozygous reference.  The output column GT represents this
genotype allele state (or unknown status) and GTcount the number of subject harboring the corresponding genotype.  The pivot commands
transform the output into multiple horizontal columns representing the different statuses for cases and controls.</p>
<p>It is also possible to perform case-control statistics for multiple phenotypes, using the same variant input stream</p>
<div class="highlight-gor notranslate"><div class="highlight"><pre><span></span>/* Creating arbitrary case control status on PNs */

<span class="nd">create</span> #phenocasecontrol# = <span class="nd">nor</span> [#PNbuckets#] | <span class="nb">SELECT</span> #<span class="na">PN</span>
| <span class="nb">CALC</span> phenotype = &#39;PhenoA,PhenoB&#39; | <span class="nb">SPLIT</span> phenotype | <span class="nb">CALC</span> ccstatus = if(random()&gt;0.5,&#39;CASE&#39;,&#39;CTRL&#39;);

<span class="nd">gor</span> [#horvars#] | <span class="nb">CSVCC</span> [#PNbuckets#] [#casecontrol#] -<span class="no">gc</span> #3,#4 -u 3
| <span class="nb">PIVOT</span> -<span class="no">gc</span> <span class="na">reference</span>,<span class="na">call</span>,pheno,ccsatus GT -v 0,1,2,3 -e 0
| <span class="nb">PIVOT</span> -<span class="no">gc</span> <span class="na">reference</span>,<span class="na">call</span>,pheno -v CASE,CTRL ccstatus -e 0
| <span class="nb">CALC</span> pVal_mm EFORM(PVAL(case_2_GTcount*2+case_1_GTcount,case_0_GTcount*2
                          + case_1_GTcount,ctrl_2_GTcount*2+ctrl_1_GTcount,ctrl_0_GTcount*2 + ctrl_1_GTcount),5,1)
| <span class="nb">SELECT</span> 1,2,<span class="na">reference</span>,<span class="na">call</span>,pheno,pVal_mm
| <span class="nb">PIVOT</span> -<span class="no">gc</span> <span class="na">reference</span>,<span class="na">call</span> pheno -v &#39;PhenoA,PhenoB&#39; -e &#39;NA&#39;
</pre></div>
</div>
<p>Below are unit tests that show equivalence queries that use GROUP count and compare it with CSVCC:</p>
<div class="highlight-gor notranslate"><div class="highlight"><pre><span></span><span class="nd">create</span> #bucket# = norrows 10 | calc bucket &#39;b1&#39; | calc <span class="na">PN</span> &#39;<span class="na">PN</span>&#39;+<span class="k">str</span>(#1) | select pn,bucket;

<span class="nd">create</span> #gt# = gorrow chr1,1 | calc <span class="na">alt</span> &#39;A&#39; | calc ref &#39;C&#39;
| merge &lt;(gorrow chr1,2 | calc <span class="na">alt</span> &#39;G&#39; | calc ref &#39;T&#39;)
| merge &lt;(gorrow chr1,3 | calc <span class="na">alt</span> &#39;A&#39; | calc ref &#39;T&#39;)
| multimap -cartesian &lt;(<span class="nd">nor</span> [#bucket#] | rownum)
| calc gt 1+mod(rownum*4+<span class="na">pos</span>,10)
| select 1,2,ref,<span class="na">alt</span>,gt,pn
| where <span class="na">PN</span> != &#39;PN0&#39;
| gtgen -<span class="no">gc</span> ref,<span class="na">alt</span> [#bucket#] &lt;(gorrow chr1,0,1 | multimap -cartesian [#bucket#] | select 1-3,pn);

<span class="nd">gor</span> [#gt#]
| csvsel -<span class="no">gc</span> ref,<span class="na">alt</span>,bucket -vs 1 [#bucket#] &lt;(<span class="nd">nor</span> [#bucket#] | select pn) -tag <span class="na">PN</span>
| group 1 -<span class="no">gc</span> ref,<span class="na">alt</span>,value -<span class="no">count</span>
| calc source &#39;group&#39;
| rename value GT | rename allcount GTcount

| merge &lt;(<span class="nd">gor</span> [#gt#]
| csvsel -<span class="no">gc</span> ref,<span class="na">alt</span>,bucket -vs 1 [#bucket#] &lt;(<span class="nd">nor</span> [#bucket#] | select pn)
| csvcc -<span class="no">gc</span> ref,<span class="na">alt</span> -vs 1 [#bucket#] &lt;(<span class="nd">nor</span> [#bucket#] | select pn | calc pheno &#39;pheno&#39;)
| calc source &#39;vs&#39;
| hide cc
)

| merge &lt;(<span class="nd">gor</span> [#gt#]
| csvsel -<span class="no">gc</span> ref,<span class="na">alt</span>,bucket -vs 1 [#bucket#] &lt;(<span class="nd">nor</span> [#bucket#] | select pn)
| replace values fsvmap(values,1,&#39;x&#39;,&#39;,&#39;)
| csvcc -<span class="no">gc</span> ref,<span class="na">alt</span> -s &#39;,&#39; [#bucket#] &lt;(<span class="nd">nor</span> [#bucket#] | select pn | calc pheno &#39;pheno&#39;)
| calc source &#39;sep&#39;
| hide cc
)
| group 1 -<span class="no">gc</span> 2-source[-1] -set -dis -sc source
| throwif dis_source != 3
</pre></div>
</div>
<p>An example showing GT values that are of different lengths:</p>
<div class="highlight-gor notranslate"><div class="highlight"><pre><span></span>/* same creates as in previous example *
<span class="nd">gor</span> [#gt#]
| csvsel -<span class="no">gc</span> ref,<span class="na">alt</span>,bucket -vs 1 [#bucket#] &lt;(<span class="nd">nor</span> [#bucket#] | select pn) -tag <span class="na">PN</span>
| replace value value*2
| group 1 -<span class="no">gc</span> ref,<span class="na">alt</span>,value -<span class="no">count</span>
| calc source &#39;group&#39;
| rename value GT | rename allcount GTcount


| merge &lt;(<span class="nd">gor</span> [#gt#]
| csvsel -<span class="no">gc</span> ref,<span class="na">alt</span>,bucket -vs 1 [#bucket#] &lt;(<span class="nd">nor</span> [#bucket#] | select pn)
| replace values fsvmap(values,1,&#39;<span class="k">int</span>(x)*2&#39;,&#39;,&#39;)
| csvcc -<span class="no">gc</span> ref,<span class="na">alt</span> -s &#39;,&#39; [#bucket#] &lt;(<span class="nd">nor</span> [#bucket#] | select pn | calc pheno &#39;pheno&#39;)
| calc source &#39;sep&#39;
| hide cc
)
| group 1 -<span class="no">gc</span> 2-source[-1] -set -dis -sc source
| throwif dis_source != 2
</pre></div>
</div>
<div class="section" id="related-commands">
<h3>Related commands<a class="headerlink" href="#related-commands" title="Permalink to this headline">¶</a></h3>
<p><a class="reference internal" href="CSVSEL.html#csvsel"><span class="std std-ref">CSVSEL</span></a></p>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2020 GOR Open Source Project

    </p>
  </div> 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>