

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>GTLD &mdash; GOR Open Source v.2.0 documentation</title>
  

  
  
    <link rel="shortcut icon" href="../_static/favicon.ico"/>
  
  
  

  
  <script type="text/javascript" src="../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../_static/css/gorpipe.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <nav class="wy-master-nav">
    <div class="gor-logo-home">
        <a href="../index.html">
            <img src="../_static/GORLogoText.svg" class="logo" />
        </a>
    </div>
    <div role="search" class="wxnc-search-form">
        <button class="wxnc-search-button" form="wxnc-search-form" formmethod="get">
            <i class="fa fa-search" aria-hidden="true"> </i>
        </button>
        <form id="wxnc-search-form" class="wy-form" action="../search.html" method="get">
            <input class="wxnc-no-right-border" size="30" type="text" name="q" placeholder="Search the user manual..."/>
            <input type="hidden" name="check_keywords" value="yes" />
            <input type="hidden" name="area" value="default" />
        </form>
    </div>
</nav>

  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            
            
              <div class="version">
                <a href="../index.html">
                  GOR Open Source
                  <p class="wxnc-subtitle">Documentation</p>
                </a>
              </div>
            
          

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../introduction.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="../commands.html">Commands</a></li>
<li class="toctree-l1"><a class="reference internal" href="../functions.html">Functions</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">GOR Open Source</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html">Docs</a> &raquo;</li>
        
      <li>GTLD</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <span style="float:right; padding-top:10px; color:#BABABA;">Used in: gor only</span><div class="section" id="gtld">
<span id="id1"></span><h1>GTLD<a class="headerlink" href="#gtld" title="Permalink to this headline">¶</a></h1>
<p>The <strong>GTLD</strong> command is used to calculate linkage disequilibrium between genotypes in different locations.
The command takes in an input stream and uses a self-join style way to calculate the LD in the surrounding region
as specified with the <code class="docutils literal notranslate"><span class="pre">-f</span></code> option.  Note that large span requires large memory buffers.  The command can be
used in parallel, however, then one must be careful that the genomic range partitions overlap by the amount
specified with the <code class="docutils literal notranslate"><span class="pre">-f</span></code> option.  Similarly, it is possibly to use partition parallelism via <strong>PARTGOR</strong>
and separate the sum step and the calculation steps for Dprime and r.</p>
<p>The input stream must have <code class="docutils literal notranslate"><span class="pre">VALUES</span></code> column storing the genotypes (as 0,1,2, or 3), with or without a <code class="docutils literal notranslate"><span class="pre">BUCKET</span></code>
column.</p>
<p>To represent the variant pairs the output columns are chrom,pos,ref,alt,distance,posx,refx,altx,
and the genotype pair counts LD_g00,LD_g10,LD_g20,LD_g01,LD_g11,LD_g21,LD_g02,LD_g12,LD_g22
and from the genotype counts the -calc option uses the EM algorithm to phase the haplotypes and estimate
LD_D,LD_Dp,LD_r,De, using the following standard formulas:</p>
<blockquote>
<div><p>Nh = h00+h01+h10+h11
p1 = (h00+h01)/Nh
q1 = (h00+h10)/Nh
LD_D = h00*h11-h10*h01
d = if (LD_D&lt;0.0) max((p1*q1),(1.0-p1)*(1.0-q1)) else min((p1*(1-q1)),(1.0-p1)*q1)
rd = sqrt(p1*(1.0-p1)*q1*(1.0-q1))
LD_Dp = LD_D/d
LD_r = LD_D/rd</p>
</div></blockquote>
<div class="section" id="usage">
<h2>Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h2>
<div class="highlight-gor notranslate"><div class="highlight"><pre><span></span><span class="nd">gor</span> ... | GTLD [ attributes ]
</pre></div>
</div>
</div>
<div class="section" id="options">
<h2>Options<a class="headerlink" href="#options" title="Permalink to this headline">¶</a></h2>
<table class="docutils align-default">
<colgroup>
<col style="width: 17%" />
<col style="width: 83%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">-f</span> <span class="pre">distance</span></code></p></td>
<td><p>The maximum span for LD calculation in base pairs.
Note very large span can require time and memory</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">-sum</span></code></p></td>
<td><p>Aggregate the correlation counts LD_g00,LD_g10,LD_g20,LD_g01,LD_g11,LD_g21,LD_g02,LD_g12 and LD_g22
from single genotype character values column.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">-calc</span></code></p></td>
<td><p>Calculate Dprime and the correlation coefficient, r, from LD_g00,..,LD_g22.</p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="examples">
<h2>Examples<a class="headerlink" href="#examples" title="Permalink to this headline">¶</a></h2>
<div class="highlight-gor notranslate"><div class="highlight"><pre><span></span>/* An example calculating LD from a stream of (chr,<span class="na">pos</span>,ref,<span class="na">alt</span>,bucket,values) */

<span class="nd">pgor</span> -split 10000000:100000 [#buckethorvars#] | GTLD -sum -calc -<span class="no">f</span> 100000
| where posx != <span class="na">pos</span> and (abs(LD_Dp) &gt; 0.2 or abs(LD_r)&gt;0.2)
| where ##WHERE_SPLIT_WINDOW##
</pre></div>
</div>
<p>The above query example calculates LD using genomic range parallelism from all the variants in all the buckets.
The following query calculate LD in a single gene:</p>
<div class="highlight-gor notranslate"><div class="highlight"><pre><span></span><span class="nd">gor</span> #genes# | where <span class="na">gene_symbol</span> = &#39;BRCA2&#39; | join -<span class="no">segsnp</span> -<span class="no">f</span> 1000000 -ir &lt;(<span class="nd">gor</span> [#buckethorvars#])
| GTLD -sum -calc -<span class="no">f</span> 100000 | where posx != <span class="na">pos</span> and (abs(LD_Dp) &gt; 0.2 or abs(LD_r)&gt;0.2)
</pre></div>
</div>
<p>Finally, the following query uses sample parallelism to calculate the LD, allowing for bigger range with less memory usage:</p>
<div class="highlight-gor notranslate"><div class="highlight"><pre><span></span><span class="nd">create</span> xxx = partgor -dict horvars.gord &lt;(<span class="nd">pgor</span> horvars.gord -<span class="no">f</span> #{tags} -nf
| csvsel -<span class="no">gc</span> <span class="na">reference</span>,<span class="na">call</span> -u 3 horvarsbuckets.tsv &lt;(<span class="nd">nor</span> horvarsbuckets.tsv | select #1 | where listhasany(<span class="na">PN</span>,&#39;#{tags}&#39;))
| varjoin -r &lt;(<span class="nd">gor</span> #VEP# | where max_impact in (&#39;HIGH&#39;,&#39;MODERATE&#39;) | select 1-4,max_consequence,max_impact)
| GTLD -sum -<span class="no">f</span> 1000000 | where posx != <span class="na">pos</span>;

<span class="nd">pgor</span> [xxx] | GTLD -calc | where (abs(LD_Dp) &gt; 0.2 or abs(LD_r)&gt;0.2) | hide LD_g00-LD_g22 | sort 1 -<span class="no">gc</span> <span class="na">reference</span>,<span class="na">call</span>,LD_r:n
</pre></div>
</div>
<p>Note that it is possible to have asymmetry in the LD calculation pairs. I.e. you can specify which variants in the input stream are only used as the left-variant in the pair.</p>
<p>An example of this is shown below:</p>
<div class="highlight-gor notranslate"><div class="highlight"><pre><span></span><span class="nd">gor</span>  varsofinterest.gor
| varjoin -ir #gtfreeze#
| varjoin -r -e 0 &lt;(<span class="nd">gor</span> myleftvars.gor | select 1-4 | calc useonlyasleftvar 1)
| GTLD -sum -calc -<span class="no">f</span> #maxlddist#
</pre></div>
</div>
<p>This scripts demonstrates how to test the caluclation of D from first principles by
simulating haplotypes of biallelic SNPs in disequilibrium and genotypes based on HWE</p>
<div class="highlight-gor notranslate"><div class="highlight"><pre><span></span><span class="nd">def</span> #pns# = 100000;
<span class="nd">def</span> #<span class="no">p</span># = 0.3;
<span class="nd">def</span> #q# = 0.02;
<span class="nd">def</span> #D# = 0.1; /* Should ideally be smaller than all of: <span class="no">p</span>*q, (1-<span class="no">p</span>)*(1-q), (1-<span class="no">p</span>)*q, <span class="no">p</span>*(1-q) */

<span class="nd">nor</span> &lt;(<span class="nd">gor</span> &lt;(norrows #pns#  | calc <span class="no">p</span> #<span class="no">p</span># | calc q #q# | calc D #D#
| calc h1 if(random()&lt;<span class="no">p</span>,if(random()&lt;(<span class="no">p</span>*q+D)/<span class="no">p</span>,&#39;0_0&#39;,&#39;0_1&#39;),if(random()&lt;((1-<span class="no">p</span>)*q-D)/(1-<span class="no">p</span>),&#39;1_0&#39;,&#39;1_1&#39;))
| calc h2 if(random()&lt;<span class="no">p</span>,if(random()&lt;(<span class="no">p</span>*q+D)/<span class="no">p</span>,&#39;0_0&#39;,&#39;0_1&#39;),if(random()&lt;((1-<span class="no">p</span>)*q-D)/(1-<span class="no">p</span>),&#39;1_0&#39;,&#39;1_1&#39;))
| colsplit h1 2 snp_f -s &#39;_&#39;
| colsplit h2 2 snp_m -s &#39;_&#39;
| calc gt1 decode(snp_f_1+&#39;_&#39;+snp_m_1,&#39;0_0,0,0_1,1,1_0,1,1_1,2&#39;)
| calc gt2 decode(snp_f_2+&#39;_&#39;+snp_m_2,&#39;0_0,0,0_1,1,1_0,1,1_1,2&#39;)
| group -lis -sc gt1,gt2 -s &#39;&#39; -len 1000000
| calc values lis_gt1+&#39;,&#39;+lis_gt2
| hide lis_*
| split values
| rownum
| rename rownum <span class="na">pos</span>
| calc chrom &#39;chr1&#39;
| calc ref &#39;A&#39;
| calc <span class="na">alt</span> &#39;T&#39;
)
| select chrom,<span class="na">pos</span>,ref,<span class="na">alt</span>,values
| gtld -sum -<span class="no">f</span> 100 | gtld -calc /* can also be written as single GTLD -sum -calc */
)
| multimap -cartesian &lt;(norrows #pns# | calc <span class="no">p</span> #<span class="no">p</span># | calc q #q# | calc D #D#
| calc h1 if(random()&lt;<span class="no">p</span>,if(random()&lt;(<span class="no">p</span>*q+D)/<span class="no">p</span>,&#39;0_0&#39;,&#39;0_1&#39;),if(random()&lt;((1-<span class="no">p</span>)*q-D)/(1-<span class="no">p</span>),&#39;1_0&#39;,&#39;1_1&#39;))
| group -<span class="no">gc</span> h1 -<span class="no">count</span>
| replace allcount <span class="k">float</span>(allcount)/#pns#
| pivot h1 -v 0_0,0_1,1_0,1_1 -e 0
| calc De #1*#4-#2*#3
| select De)
| select <span class="na">distance</span>,LD_D-LD_r,De
| throwif <span class="na">distance</span> = 0 and abs(ld_Dp-1.0)&gt;0.01
| throwif abs(<span class="na">distance</span>) = 1 and abs(LD_D - De)&gt;0.01
</pre></div>
</div>
<div class="section" id="related-commands">
<h3>Related commands<a class="headerlink" href="#related-commands" title="Permalink to this headline">¶</a></h3>
<p><a class="reference internal" href="CSVCC.html#csvcc"><span class="std std-ref">CSVCC</span></a> <a class="reference internal" href="CSVSEL.html#csvsel"><span class="std std-ref">CSVSEL</span></a> <a class="reference internal" href="GTGEN.html#gtgen"><span class="std std-ref">GTGEN</span></a></p>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2020 GOR Open Source Project

    </p>
  </div> 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>