

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>KING &mdash; GOR Open Source v.2.0 documentation</title>
  

  
  
    <link rel="shortcut icon" href="../_static/favicon.ico"/>
  
  
  

  
  <script type="text/javascript" src="../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script type="text/javascript" src="../_static/jquery.js"></script>
        <script type="text/javascript" src="../_static/underscore.js"></script>
        <script type="text/javascript" src="../_static/doctools.js"></script>
        <script type="text/javascript" src="../_static/language_data.js"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../_static/css/gorpipe.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <nav class="wy-master-nav">
    <div class="gor-logo-home">
        <a href="../index.html">
            <img src="../_static/GORLogoText.svg" class="logo" />
        </a>
    </div>
    <div role="search" class="wxnc-search-form">
        <button class="wxnc-search-button" form="wxnc-search-form" formmethod="get">
            <i class="fa fa-search" aria-hidden="true"> </i>
        </button>
        <form id="wxnc-search-form" class="wy-form" action="../search.html" method="get">
            <input class="wxnc-no-right-border" size="30" type="text" name="q" placeholder="Search the user manual..."/>
            <input type="hidden" name="check_keywords" value="yes" />
            <input type="hidden" name="area" value="default" />
        </form>
    </div>
</nav>

  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            
            
              <div class="version">
                <a href="../index.html">
                  GOR Open Source
                  <p class="wxnc-subtitle">Documentation</p>
                </a>
              </div>
            
          

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../introduction.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="../commands.html">Commands</a></li>
<li class="toctree-l1"><a class="reference internal" href="../functions.html">Functions</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">GOR Open Source</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html">Docs</a> &raquo;</li>
        
      <li>KING</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <span style="float:right; padding-top:10px; color:#BABABA;">Used in: gor only</span><section id="king">
<span id="id1"></span><h1>KING<a class="headerlink" href="#king" title="Permalink to this headline">¶</a></h1>
<p>The <strong>KING</strong> command is used to calculate parameters that can be used in the KING algorithm (see Ani Manichaikul et.al.
Bioinformatics, Vol. 26 no. 22 2010, pages 2867–2873).</p>
<p>The left-input stream is of the same type as for <strong>CSVSEL</strong>, i.e. it must contain the columns <code class="docutils literal notranslate"><span class="pre">values</span></code> and <code class="docutils literal notranslate"><span class="pre">bucket</span></code>.
The tag-bucket relation is used to define how the values for different tags are stored in the value column in each bucket.
The tag selections specify which tags are selected and the calculation is carried out for each tag (PN) in the first
tag list against each tag in the second tag list, for every variant in the left-input stream.  The left-input stream
must include a column with the allele frequency, named AF.  Such column is most easily added to the horizontal genotype
variant format with a <strong>VARJOIN</strong>.</p>
<p>The output is a row for each tag-pair, using columns named PN1 and PN2.  Since the output is an aggregate over the genome
is also includes dummy columns for Chrom and Pos with the values “ChrA” and 0, respectively.  The important columns are
however PN1, PN2, IBS0, XX, tpq, kpq, Nhet, Nhom, NAai, NAaj, count, pi0, phi, and theta.  See the code examples
and the KING paper for further details related to the <strong>KING</strong> command.</p>
<section id="usage">
<h2>Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h2>
<div class="highlight-gor notranslate"><div class="highlight"><pre><span></span><span class="nd">gor</span> ... | KING tagbucketrelation tagselection1 tagselection2  [-<span class="no">gc</span> cols] [(-s sep | -vs charsize)]
</pre></div>
</div>
</section>
<section id="options">
<h2>Options<a class="headerlink" href="#options" title="Permalink to this headline">¶</a></h2>
<table class="docutils align-default">
<colgroup>
<col style="width: 17%" />
<col style="width: 83%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">-gc</span> <span class="pre">cols</span></code></p></td>
<td><p>Grouping columns other than (chrom,pos).  NOTE, one of the must include AF</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">-s</span> <span class="pre">sep</span></code></p></td>
<td><p>Specify separator for the elements in values.  Default to comma.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">-vs</span> <span class="pre">size</span></code></p></td>
<td><p>Use a fixed character size for values, e.g. rather than variable length separated with comma.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">-pi0thr</span> <span class="pre">value</span></code></p></td>
<td><p>Threshold value to filter output on pi0, e.g. to pass we must have pi0 &lt; value</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">-phithr</span> <span class="pre">value</span></code></p></td>
<td><p>Threshold value to filter output on phi, e.g. to pass we must have phi &gt; value</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">-thetathr</span> <span class="pre">value</span></code></p></td>
<td><p>Threshold value to filter output on theta, e.g. to pass we must have theta &gt; value</p></td>
</tr>
</tbody>
</table>
</section>
<section id="examples">
<h2>Examples<a class="headerlink" href="#examples" title="Permalink to this headline">¶</a></h2>
<div class="highlight-gor notranslate"><div class="highlight"><pre><span></span>The following query shows an example where the the QUEEN command is used to estimate which sample pairs
to evaluate in a KING2 analysis.  The QUEEN command uses two <span class="na">PN</span> lists that are generated by grouping the
pn-pairs according to the bucket groups of the genotype matrix into an upper-triagonal matrix.
Only the genotype buckets necessary for evaluating the pairs are read from disk in each job and following the analysis,
the output-pairs are filtered to avoid storing symmetric pairs.

<span class="nd">def</span> #buckets# = freezes/ukbb_500k/imputed_comm_vep95/buckets.tsv;
<span class="nd">def</span> #hgt# = freezes/ukbb_500k/imputed_comm_vep95/variants.gord;
<span class="nd">def</span> #af# = freezes/ukbb_500k/imputed_comm_vep95/metadata/AF.gorz;

<span class="nd">create</span> #useaf# = <span class="nd">gor</span> #af# /* Only use rare variants for Queen */
| where len(ref)=len(<span class="na">alt</span>) and af &lt; 0.0025
| select 1-4,af;

<span class="nd">create</span> #buckgroups# = <span class="nd">nor</span> #buckets#
| select bucket
| distinct
| sort -c bucket:n
| rownum
| calc gr div(rownum,4);  /* larger groups mean more memory usage but less repeated IO access of genotypes */

<span class="nd">create</span> #parts# = <span class="nd">nor</span> [#buckgroups#]
| select gr
| distinct
| multimap -cartesian &lt;(<span class="nd">nor</span> [#buckgroups#] | select gr | distinct)
| where gr &lt;= grx
| rownum
| calc <span class="no">p</span> 1+mod(rownum,2) /* Allow us to avid limit on the number of maximum job per query, e.g. 1000 */
| hide rownum;

<span class="nd">create</span> #sharepairs1# = parallel -parts &lt;(<span class="nd">nor</span> [#parts#] | where <span class="no">p</span> = 1)
       &lt;(<span class="nd">nor</span> &lt;(<span class="nd">gor</span> #hgt# -nf -ff &lt;(<span class="nd">nor</span> #buckets# | inset -c bucket &lt;(<span class="nd">nor</span> [#buckgroups#] | where gr = &#39;#{col:gr}&#39; or gr = &#39;#{col:grx}&#39;))
       | where chrom &lt; &#39;chrX&#39; | varjoin -i &lt;(<span class="nd">gor</span> [#useaf#])
       | QUEEN #buckets#  &lt;(<span class="nd">nor</span> #buckets# | inset -c bucket &lt;(<span class="nd">nor</span> [#buckgroups#] | where gr = &#39;#{col:gr}&#39;) | select pn)
                  &lt;(<span class="nd">nor</span> #buckets# | inset -c bucket &lt;(<span class="nd">nor</span> [#buckgroups#] | where gr = &#39;#{col:grx}&#39;) | select pn) -vs 1 -minSharing 0.02
       | multimap -c pn1 #buckets# | multimap -c pn2 #buckets# | where bucket &lt; bucketx or bucket = bucketx and pn1 &lt; pn2 | hide bucket,bucketx )
       | select pn1- );

<span class="nd">create</span> #sharepairs2# = parallel -parts &lt;(<span class="nd">nor</span> [#parts#])
       &lt;(<span class="nd">nor</span> &lt;(<span class="nd">gor</span> #hgt# -nf -ff &lt;(<span class="nd">nor</span> #buckets# | inset -c bucket &lt;(<span class="nd">nor</span> [#buckgroups#] | where gr = &#39;#{col:gr}&#39; or gr = &#39;#{col:grx}&#39;))
       | where chrom &lt; &#39;chrX&#39; | varjoin -i &lt;(<span class="nd">gor</span> [#useaf#])
       | QUEEN #buckets#  &lt;(<span class="nd">nor</span> #buckets# | inset -c bucket &lt;(<span class="nd">nor</span> [#buckgroups#] | where gr = &#39;#{col:gr}&#39;) | select pn)
                  &lt;(<span class="nd">nor</span> #buckets# | inset -c bucket &lt;(<span class="nd">nor</span> [#buckgroups#] | where gr = &#39;#{col:grx}&#39;) | select pn) -vs 1 -minSharing 0.02
       | multimap -c pn1 #buckets# | multimap -c pn2 #buckets# | where bucket &lt; bucketx or bucket = bucketx and pn1 &lt; pn2 | hide bucket,bucketx )
       | select pn1- );


<span class="nd">create</span> #usedvars# = <span class="nd">pgor</span> #af#
| where len(ref)=len(<span class="na">alt</span>)
| select 1-4,af
| where abs(af-0.5)&lt;0.4
| where random()&lt;1e-2
| granno 10000 -<span class="no">min</span> -<span class="no">ic</span> <span class="na">pos</span>
| where min_pos = <span class="na">pos</span>
| hide min_pos;

<span class="nd">create</span> #af# = <span class="nd">pgor</span> [#usedvars#]
| select 1-4,af
| where len(#3)=len(#4)
| select 1-4,af
| calc tpq 2*af*af*(1.0-af)*(1.0-af)
| calc kpq 2.0*af*(1-af);

<span class="nd">create</span> #varsharing# = parallel -parts &lt;(<span class="nd">nor</span> [#parts#] )
&lt;(<span class="nd">nor</span> &lt;(<span class="nd">gor</span> #hgt# -nf -ff &lt;(<span class="nd">nor</span> #buckets# | inset -c bucket &lt;(<span class="nd">nor</span> [#buckgroups#] | where gr = &#39;#{col:gr}&#39; or gr = &#39;#{col:grx}&#39;))
| where len(ref)=len(<span class="na">alt</span>) | until chrom = &#39;chrX&#39; | varjoin -r [#af#]
| KING2 -<span class="no">gc</span> af #buckets# &lt;(<span class="nd">nor</span> [#sharepairs1#] | merge [#sharepairs2#]
                            /* unfortunately we don&#39;t have partitioned access to this data, hence the filtering with <span class="nb">INSET</span> */
                          | inset -c pn1 &lt;(<span class="nd">nor</span> #buckets# | inset -c bucket &lt;(<span class="nd">nor</span> [#buckgroups#] | where gr = &#39;#{col:gr}&#39; ))
                          | inset -c pn2 &lt;(<span class="nd">nor</span> #buckets# | inset -c bucket &lt;(<span class="nd">nor</span> [#buckgroups#] | where gr = &#39;#{col:grx}&#39; )) | select pn1,pn2 )
                          -vs 1 -pi0thr 0.76 -phithr 0.039   )
| group -<span class="no">gc</span> pn1,pn2 -sum -fc IBS0-
| calc pi0 sum_IBS0/<span class="k">float</span>(sum_tpq)
| calc phi 0.5-<span class="k">float</span>(sum_XX)/<span class="k">float</span>(4.0*sum_kpq)
| calc theta (sum_Nhet-2.0*sum_Nhom)/(sum_NAai+sum_NAaj)
);

<span class="nd">def</span> expo($1,$2,$3) = exp(-sqr( $1 - $2 )/<span class="k">float</span>(2* $3 * $3 )) / $3;

<span class="nd">nor</span> [#varsharing#]
| calc monozygotic if(phi &gt; pow(2.0,-1.5) and phi &lt; 0.1,1,0)
| calc parent_offspring if(phi &gt; pow(2.0,-2.5) and phi &lt; pow(2.0,-1.5) and pi0 &lt; 0.1,1,0)
| calc full_sib if(phi &gt; pow(2.0,-2.5) and phi &lt; pow(2.0,-1.5) and pi0 &gt; 0.1  and pi0 &lt; 0.365,1,0)
| calc second_degree if(phi &gt; pow(2.0,-3.5) and phi &lt; pow(2.0,-2.5) and pi0 &gt; 0.365 and pi0 &lt; 1.0-pow(2,-1.5),1,0)
| calc third_degree if(phi &gt; pow(2.0,-4.5) and phi &lt; pow(2.0,-3.5) and pi0 &gt; 1.0-pow(2,-1.5) and pi0 &lt; 1.0-pow(2,-2.5),1,0)
| calc temp if(monozygotic=1,&#39;Monozygotic twin,&#39;,&#39;&#39;)+if(parent_offspring=1,&#39;Parent-offspring,&#39;,&#39;&#39;)+if(full_sib=1,&#39;Full sib,&#39;,&#39;&#39;)+if(second_degree=1,&#39;2nd Degree,&#39;,&#39;&#39;)+if(third_degree=1,&#39;3rd Degree,&#39;,&#39;&#39;)
| calc relationship <span class="k">listfilter</span>(temp,&#39;len(x)&gt;1&#39;)
| calc monozygotic_dist ln( expo(pi0,0.0,0.000001) * expo(phi,0.498,0.0013) )
| calc parent_offspring_dist ln( expo(pi0,0.00297,0.0025) * expo(phi,0.248,0.01) )
| calc full_sib_dist ln( expo(pi0,0.248,0.0406) * expo(phi,0.248,0.0187) )
| calc second_degree_dist ln( expo(pi0,0.4954,0.0608) * expo(phi,0.127,0.027) )
| calc third_degree_dist ln( expo(pi0,0.745,0.0428) * expo(phi,0.059,0.02) )
| calc rel &#39;Monozygotic twin,Parent-offspring,Full sib,2nd Degree,3rd Degree&#39;
| calc relp <span class="k">str</span>(monozygotic_dist)+&#39;,&#39;+<span class="k">str</span>(parent_offspring_dist)+&#39;,&#39;+<span class="k">str</span>(full_sib_dist)+&#39;,&#39;+<span class="k">str</span>(second_degree_dist)+&#39;,&#39;+<span class="k">str</span>(third_degree_dist)
| calc relmax listnummax(relp)
| calc GaussRel listzipfilter(rel,relp,&#39;abs(<span class="k">float</span>(x)-relmax)&lt;0.001&#39;)
| select pn1,pn2,relationship,GaussRel,pi0,phi,theta
</pre></div>
</div>
<section id="related-commands">
<h3>Related commands<a class="headerlink" href="#related-commands" title="Permalink to this headline">¶</a></h3>
<p><a class="reference internal" href="KING2.html#king2"><span class="std std-ref">KING2</span></a> <a class="reference internal" href="#king"><span class="std std-ref">KING</span></a> <a class="reference internal" href="CSVCC.html#csvcc"><span class="std std-ref">CSVCC</span></a> <a class="reference internal" href="GTGEN.html#gtgen"><span class="std std-ref">GTGEN</span></a> <a class="reference internal" href="GTLD.html#gtld"><span class="std std-ref">GTLD</span></a></p>
</section>
</section>
</section>


           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2020 GOR Open Source Project

    </p>
  </div> 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
    <!-- Theme Analytics -->
    <script>
    (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
      (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
      m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
    })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

    ga('create', 'UA-136552918-1', 'auto');
    ga('send', 'pageview');
    </script>

    
   

</body>
</html>